File: lang-java/dream/深度优先搜索/No236.java
package 深度优先搜索;

import 默认模板.TreeNode;

/**
 * @author Matthew Han
 * @version 1.0
 * @description
 * @since 2020/12/30 14:37
 **/
public class No236 {

    TreeNode res;
    boolean l = false;
    boolean r = false;

    /**
     * #236 二叉树的最近公共祖先
     * 执行用时： 1555 ms , 在所有 Java 提交中击败了 5.02% 的用户
     * 内存消耗： 40.4 MB , 在所有 Java 提交中击败了 79.81% 的用户
     *
     * @param root
     * @param p
     * @param q
     * @return
     */
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        dfs(root, p, q);
        return res;
    }

    public void check(TreeNode curr, TreeNode p, TreeNode q) {
        if (curr == null) {
            return;
        }
        if (curr == p) {
            l = true;
        }
        if (curr == q) {
            r = true;
        }
        check(curr.left, p, q);
        check(curr.right, p, q);
    }

    public void dfs(TreeNode curr, TreeNode p, TreeNode q) {
        if (curr == null) {
            return;
        }
        l = false;
        r = false;
        check(curr, p, q);
        if (l && r) {
            res = curr;
            return;
        }
        dfs(curr.left, p, q);
        dfs(curr.right, p, q);
    }
}

File: lang-py/No236.py
"""
  author <a href="https://github.com/Matthew-Han">Matthew Han</a>
  since 2024/2/18 09:37:46
"""
from typing import List

from TreeNode import TreeNode


class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def dfs(root: 'TreeNode', t: 'TreeNode') -> List['TreeNode']:
            if root is None:
                return []
            else:
                r1 = dfs(root.left, t)
                r2 = dfs(root.right, t)
                ans = r1 if r1 else r2
                if root.val == t.val or len(r1) > 0 or len(r2) > 0:
                    ans.append(root)
                return ans

        path1 = list(reversed(dfs(root, p)))
        path2 = list(reversed(dfs(root, q)))
        for i in range(min(len(path1), len(path2))):
            if path1[i].val != path2[i].val:
                return path1[i - 1]
            if path1[i] == p.val or path1[i] == q:
                return path1[i]
            elif path2[i] == p or path2[i] == q:
                return path2[i]
        return root
